upgrade/src/main/java/org/odk/collect/upgrade/AppUpgrader.kt

```diff
@@ -0,0 +1,50 @@
+package org.odk.collect.upgrade
+
+import android.app.Application
+import org.odk.collect.shared.Settings
+
+class AppUpgrader(
+    private val settings: Settings,
+    private val launchState: LaunchState,
+    private val upgrades: List<Upgrade>
+) {
+    constructor(
+        key: String,
+        settings: Settings,
+        versionCode: Int,
+        installDetector: InstallDetector,
+        upgrades: List<Upgrade>
+    ) : this(
+        settings,
+        VersionCodeLaunchState(key, settings, versionCode, installDetector),
```

Comment:
Wouldn't it be better to pass it as a parameter then you wouldn't need this secondary constructor at all.

Reply:
The idea here is that we can keep `LaunchState` and `VersionCodeLaunchState` internal to `upgrade` - dependent modules don't need to know anything about them and can just pass in an `InstallDetector` etc. I intended to make `LaunchState` etc `internal` to make use of this but completely forgot. I've taken care of that now, so it's hopefully clear what the benefit of the secondary "convenience" constructor is.

Reply:
I see that makes sense.